File: question/neetcode150/stack/daily-temperatures.py
from typing import List
from collections import deque

class Solution:
    def dailyTemperatures(self, a: List[int]) -> List[int]:
        ans = [0]* len(a)
        stack = []

        for i,n in enumerate(a):
            while stack and stack[-1][1] < n:
                k,v = stack.pop()
                ans[k] = i-k
            stack.append((i,n))

        return ans

File: question/neetcode150/stack/daily-temperatures.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& a) {
        int n = a.size();
        stack<pair<int,int>> s;
        vector<int> ans(n);

        for(int i=0;i<n;i++){
            while(!s.empty() and s.top().second<a[i]){
                auto [k,v] = s.top();
                s.pop();
                ans[k] = i-k;
            }
            s.push({i,a[i]});
        }

        return ans;
    }
};